### DISCUSSION TITLE
R._next predicate

### DISCUSSION THREAD

**Discussion by voidpetal** [LukasZahradnik/PyNeuraLogic#48] [General]
Hello,

I have been using the `R.special.next(V.X, V.Y) predicate` in my templates, but then I saw in the following notebook that I have to define the `R._next` predicate for each pair of integers I will be using. My question is whether this step is necessary or somehow implicit, because I don't see it in your implementation of RNNs, for example.

https://github.com/LukasZahradnik/PyNeuraLogic/blob/master/examples/RecursiveXORGeneralization.ipynb

Thank you!

**Comment by LukasZahradnik** [comment]
Hello,

Those are two different predicates used for the same use case. The example you linked predates the **special** next predicate, so it uses an ordinary predicate to define a sequence of integers. Because this functionality was needed quite often, we have added the **special** `next` predicate  that kinda simplifies that. Thanks for bringing this up. I will update the example.

Regarding your question, when using the special next predicate, the sequence of integers is implicit, but it only works for an already defined set of integers.

For example:
```
R.get_next(V.X) <= R.special.next(1, V.X)
```
If your dataset/template/query contains somewhere in terms number _2_ (e.g., `R.feature(2)`), then you can get grounded `R.get_next(2)`.
If there is no _2_ anywhere in your dataset/template/query, then the variable `V.X` has no substitution.
So basically, the predicate will not "produce" an unseen term (integer).

**Reply by GustikS** [reply]
I think that there might be a space for another convenience sugar that would actually create those numeric constants, something like `R.special.range(0,10)` placed anywhere in the template, just producing `num(0), num(1), ...` as a quick hack, maybe? :)

**Reply by LukasZahradnik** [reply]
The convenience sugar here is Python's range:
```py
template += R.num(i) for i in range(0, 10)
```

I would like to avoid all those non-transparent stuff - taking one relation and transforming it into N completely different relations.
Maybe we could have a module for that... 
```py
from  neuralogic.nn.module import Range

template += Range("num", 0, 10)
```

**Reply by GustikS** [reply]
:D I agree, probably just adding this to the documentation on the special next/2 predicate will be the best solution...

**Reply by voidpetal** [reply]
So, as an example, if my dataset contains vertices labeled 1, 2, 3, ..., then the `R.special.next(1, V.X)` will ground and vertex 2 will be "after" vertex 1. Otherwise, it will not ground at all?

**Reply by LukasZahradnik** [reply]
@erhc Yes, correct. It has to contain the constant number 2 in terms somewhere. It does not matter where.

**Reply by voidpetal** [reply]
Now I am seeing some weird results. With the addition of these new facts, my model performs with a significantly higher error, both on train and test datasets.

**Reply by GustikS** [reply]
@erhc then you should just check carefully on some small example what your model was actually doing before, because without those numeric constants, it was very likely degenerating into something much simpler (which might be a good thing for learning).
